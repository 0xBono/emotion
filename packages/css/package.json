{
  "name": "@emotion/css",
  "version": "0.9.8",
  "description": "a function to serialize css and object styless",
  "main": "dist/css.cjs.js",
  "module": "dist/css.esm.js",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/emotion/tree/master/next-packages/css",
  "dependencies": {
    "@emotion/serialize": "^0.9.1",
    "@emotion/utils": "^0.8.2",
    "babel-plugin-emotion": "^9.2.10"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist",
    "macro.js"
  ],
  "umd:main": "./dist/index.min.js",
  "browser": {
    "./dist/css.cjs.js": "./dist/css.browser.cjs.js",
    "./dist/css.esm.js": "./dist/css.browser.esm.js"
  }
}
