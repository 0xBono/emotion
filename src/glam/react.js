import React from 'react'

export default function glam (tag, className) {
  return class Target extends React.Component {
    render () {
      return React.createElement(tag, { className })
    }
  }
}
