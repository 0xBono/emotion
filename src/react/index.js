import { Component, createElement as h } from 'react'
import PropTypes from 'prop-types'
import map from '@arr/map'
import { css } from '../index'
import { omit } from '../utils'
import { CHANNEL } from './constants'

export {
  flush,
  css,
  injectGlobal,
  fontFace,
  keyframes,
  hydrate,
  objStyle
} from '../index'

export default function (tag, objs, vars, content) {
  if (!tag) {
    throw new Error(
      'You are trying to create a styled element with an undefined component.\nYou may have forgotten to import it.'
    )
  }

  const componentTag = tag.displayName || tag.name || 'Component'

  class Styled extends Component {
    state = {
      theme: {}
    }

    componentWillMount () {
      if (this.context[CHANNEL]) {
        this.setState({ theme: this.context[CHANNEL].getState() })
      }
    }

    componentDidMount () {
      if (this.context[CHANNEL]) {
        this.unsubscribe = this.context[CHANNEL].subscribe(this.setTheme)
      }
    }

    componentWillUnmount () {
      if (typeof this.unsubscribe === 'function') {
        this.unsubscribe()
      }
    }

    render () {
      const { props, state } = this
      const mergedProps = {
        ...props,
        theme: state.theme
      }

      const getValue = v => {
        return v && typeof v === 'function'
          ? (v.__emotion_spec &&
              css(
                map(v.__emotion_spec.objs, getValue),
                map(v.__emotion_spec.vars, getValue),
                v.__emotion_spec.content
              )) ||
              v(mergedProps)
          : v
      }

      const className = `${css(
        map(objs.concat(mergedProps.className.split(' ')), getValue),
        map(vars, getValue),
        content
      )}`

      return h(
        tag,
        omit(
          Object.assign({}, mergedProps, {
            ref: mergedProps.innerRef,
            className
          }),
          ['innerRef', 'theme']
        )
      )
    }

    setTheme = theme => this.setState({ theme })
  }

  Styled.contextTypes = {
    [CHANNEL]: PropTypes.object
  }

  Styled.displayName = `styled(${componentTag})`
  Styled.__emotion_spec = {
    vars,
    content,
    objs
  }
  return Styled
}
