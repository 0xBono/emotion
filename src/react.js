import React from 'react'
import css from 'glam'

export const emotion = (tag, [cls, vars = []]) => props =>
  React.createElement(
    tag,
    Object.assign({}, props, {
      className: (props.className || '') +
        ' ' +
        css(
          cls,
          vars.map(v => (v && typeof v === 'function' ? v(props) : v))
        )
    })
  )
