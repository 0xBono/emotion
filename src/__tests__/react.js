/* eslint-disable jsx-quotes,no-useless-escape,no-template-curly-in-string */
/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import css, {fragment} from 'glam'
import {emotion} from '../react'
import plugin from '../babel'

const babel = require('babel-core')

describe('glam react', () => {
  describe('babel glam component', () => {
    test('basic', () => {
      const basic = 'emotion.h1\`font-size: \$\{fontSize\}px;\`'
      const {code} = babel.transform(basic, {plugins: [plugin, 'glam/babel']})
      expect(code).toMatchSnapshot()
    })

    test('function call', () => {
      const basic = 'emotion(MyComponent)\`font-size: \$\{fontSize\}px;\`'
      const {code} = babel.transform(basic, {plugins: [plugin, 'glam/babel']})
      expect(code).toMatchSnapshot()
    })
  })

  test('basic render', () => {
    const fontSize = 20
    const H1 = emotion.h1`
      font-size: ${fontSize}px;
    `

    const tree = renderer
      .create(
        <H1>
          hello world
        </H1>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('call expression', () => {
    const fontSize = 20
    const H1 = emotion('h1')`
      font-size: ${fontSize}px;
    `

    const tree = renderer
      .create(
        <H1 className={'legacy__class'}>
          hello world
        </H1>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('composition', () => {
    const fontSize = 20
    const H1 = emotion('h1')`
      font-size: ${fontSize}px;
    `

    const H2 = emotion(H1)`font-size: ${fontSize * 2 / 3}`

    const tree = renderer
      .create(
        <H2 className={'legacy__class'}>
          hello world
        </H2>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('function in expression', () => {
    const fontSize = 20
    const H1 = emotion('h1')`
      font-size: ${fontSize}px;
    `

    const H2 = emotion(H1)`font-size: ${({scale}) => fontSize * scale}`

    const tree = renderer
      .create(
        <H2 scale={2} className={'legacy__class'}>
          hello world
        </H2>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('higher order component', () => {
    const fontSize = 20
    const Content = emotion('div')`
      font-size: ${fontSize}px;
    `

    const flexColumn = Component => {
      const NewComponent = emotion(Component)`
        flex-direction: column;
      `
      NewComponent.displayName = `flexColumn${Component.displayName}`

      return NewComponent
    }

    const ColumnContent = flexColumn(Content)

    expect(ColumnContent.displayName).toBe('flexColumnundefined')

    const tree = renderer
      .create(
        <ColumnContent>
          <div
            css={`height: 100px; width: 100px; background-color: #20c997; margin: 8ch;`}
          />
          <div
            css={`height: 100px; width: 100px; background-color: #20c997; margin: 8ch;`}
          />
          <div
            css={`height: 100px; width: 100px; background-color: #20c997; margin: 8ch;`}
          />
          <div
            css={`height: 100px; width: 100px; background-color: #20c997; margin: 8ch;`}
          />
        </ColumnContent>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
