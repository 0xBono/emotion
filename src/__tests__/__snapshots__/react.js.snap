// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`glam react babel glam component basic 1`] = `"emotion(\\"h1\\", ['css-8xpzga', [fontSize]]);"`;

exports[`glam react babel glam component function call 1`] = `"emotion(MyComponent, ['css-8xpzga', [fontSize]]);"`;

exports[`glam react basic render 1`] = `
<h1
  className=" css-13wdnau vars-nujrf4"
>
  hello world
</h1>
`;

exports[`glam react call expression 1`] = `
<h1
  className="legacy__class css-13wdnau vars-nujrf4"
>
  hello world
</h1>
`;

exports[`glam react composition 1`] = `
<h1
  className="legacy__class css-vxb7tq vars-3na0zv css-13wdnau vars-nujrf4"
>
  hello world
</h1>
`;

exports[`glam react function in expression 1`] = `
<h1
  className="legacy__class css-vxb7tq vars-prycnx css-13wdnau vars-nujrf4"
  scale={2}
>
  hello world
</h1>
`;

exports[`glam react higher order component 1`] = `
<div
  className=" css-o7j1h0 css-13wdnau vars-nujrf4"
>
  <div
    className="css-1l3rx1p"
  />
  <div
    className="css-1l3rx1p"
  />
  <div
    className="css-1l3rx1p"
  />
  <div
    className="css-1l3rx1p"
  />
</div>
`;
